File: apps/visualizer/backend/api/images/stacks.py
"""Burst stack members and mutations."""

from __future__ import annotations

from flask import Blueprint, jsonify, request
from spectree import Response
from utils.db import with_db
from utils.responses import error_bad_request, error_not_found, error_server_error

from api.openapi import spec
from api.schemas.jobs import ErrorBody
from api.schemas.stacks import (
    StackMembersResponse,
    StackMergeRequest,
    StackMergeResponse,
    StackRepresentativeRequest,
    StackRepresentativeResponse,
    StackSplitMemberRequest,
    StackSplitMemberResponse,
    StackSuggestionAcceptResponse,
    StackSuggestionPairRequest,
    StackSuggestionRejectResponse,
    StackSuggestionsResponse,
)
from lightroom_tagger.core.database import (
    StackMutationError,
    library_write,
    list_pending_stack_suggestions,
    list_stack_member_keys,
    query_catalog_images_by_keys,
    reject_catalog_similarity_pair,
    stack_accept_suggestion_pair,
    stack_exists,
    stack_merge_into,
    stack_set_representative,
    stack_split_member_out,
)

from .catalog import _query_catalog_rows_for_stack_member_keys, _rows_to_catalog_api_images

stacks_bp = Blueprint("images_stacks", __name__)


def _clamp_pagination(limit: int, offset: int) -> tuple[int, int]:
    return max(1, min(int(limit), 100)), max(0, int(offset))


@stacks_bp.route("/suggestions", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=StackSuggestionsResponse),
    tags=['images-stacks'],
)
def list_stack_suggestions(db):
    """Pending catalog-similarity pairs ranked as stacks to confirm."""
    try:
        limit, offset = _clamp_pagination(
            request.args.get("limit", 20, type=int),
            request.args.get("offset", 0, type=int),
        )
        rows, total = list_pending_stack_suggestions(db, limit=limit, offset=offset)
        items: list[dict] = []
        for row in rows:
            keys = [str(row["seed_key"]), str(row["candidate_key"])]
            catalog_rows = query_catalog_images_by_keys(db, keys)
            images = _rows_to_catalog_api_images(catalog_rows)
            by_key = {img["key"]: img for img in images}
            image_a = by_key.get(str(row["seed_key"]))
            image_b = by_key.get(str(row["candidate_key"]))
            if not image_a or not image_b:
                continue
            image_a["thumbnail_url"] = f"/api/images/catalog/{image_a['key']}/thumbnail"
            image_b["thumbnail_url"] = f"/api/images/catalog/{image_b['key']}/thumbnail"
            gap = row.get("time_gap_seconds")
            items.append(
                {
                    "group_id": int(row["group_id"]),
                    "image_a": image_a,
                    "image_b": image_b,
                    "similarity": float(row["similarity"] or 0.0),
                    "why_matched": str(row["why_matched"] or ""),
                    "time_gap_seconds": int(gap) if gap is not None else None,
                }
            )
        return jsonify({"items": items, "total": total})
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/suggestions/accept", methods=["POST"])
@with_db
@spec.validate(
    json=StackSuggestionPairRequest,
    resp=Response(
        HTTP_200=StackSuggestionAcceptResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-stacks'],
)
def post_stack_suggestion_accept(db):
    """Accept a suggested pair by creating, extending, or merging stacks."""
    try:
        body = request.get_json(silent=True)
        if not body or not isinstance(body, dict):
            return error_bad_request("JSON body required")
        key_a = body.get("image_key_a")
        key_b = body.get("image_key_b")
        if not key_a or not isinstance(key_a, str) or not key_b or not isinstance(key_b, str):
            return error_bad_request("image_key_a and image_key_b required")
        with library_write(db):
            result = stack_accept_suggestion_pair(db, key_a.strip(), key_b.strip())
        # stack_accept_suggestion_pair may merge two stacks and return merged_stack_id;
        # the accept contract is stack-only (extra='forbid'), so shape to {stack}.
        return jsonify({"stack": result["stack"]}), 200
    except StackMutationError as e:
        if e.status_code == 404:
            return error_not_found("image")
        if e.status_code >= 500:
            return error_server_error(str(e))
        return error_bad_request(str(e))
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/suggestions/reject", methods=["POST"])
@with_db
@spec.validate(
    json=StackSuggestionPairRequest,
    resp=Response(HTTP_200=StackSuggestionRejectResponse, HTTP_400=ErrorBody),
    tags=['images-stacks'],
)
def post_stack_suggestion_reject(db):
    """Reject a suggested pair so it does not return on the next batch run."""
    try:
        body = request.get_json(silent=True)
        if not body or not isinstance(body, dict):
            return error_bad_request("JSON body required")
        key_a = body.get("image_key_a")
        key_b = body.get("image_key_b")
        if not key_a or not isinstance(key_a, str) or not key_b or not isinstance(key_b, str):
            return error_bad_request("image_key_a and image_key_b required")
        a = key_a.strip()
        b = key_b.strip()
        if a == b:
            return error_bad_request("image_key_a and image_key_b must differ")
        with library_write(db):
            reject_catalog_similarity_pair(db, a, b)
        return jsonify({"image_key_a": a, "image_key_b": b, "rejected": True}), 200
    except ValueError as e:
        return error_bad_request(str(e))
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/<int:stack_id>/members", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=StackMembersResponse, HTTP_404=ErrorBody),
    tags=['images-stacks'],
)
def get_stack_members(db, stack_id: int):
    """Members of a burst stack as catalog-shaped rows (representative + collapsed rules)."""
    try:
        if stack_id < 1:
            return error_not_found("stack")
        if not stack_exists(db, stack_id):
            return error_not_found("stack")

        keys = list_stack_member_keys(db, stack_id)
        if not keys:
            return jsonify({"items": []})

        catalog_rows = _query_catalog_rows_for_stack_member_keys(db, keys, score_perspective=None)
        items = _rows_to_catalog_api_images(catalog_rows)
        for it in items:
            it["thumbnail_url"] = f"/api/images/catalog/{it['key']}/thumbnail"
        return jsonify({"items": items})
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/<int:stack_id>/split-member", methods=["POST"])
@with_db
@spec.validate(
    json=StackSplitMemberRequest,
    resp=Response(
        HTTP_200=StackSplitMemberResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-stacks'],
)
def post_stack_split_member(db, stack_id: int):
    """Remove a member from a stack (solo image) or dissolve a two-member stack."""
    try:
        if stack_id < 1:
            return error_not_found("stack")
        body = request.get_json(silent=True)
        if not body or not isinstance(body, dict):
            return error_bad_request("JSON body required")
        image_key = body.get("image_key")
        if not image_key or not isinstance(image_key, str):
            return error_bad_request("image_key required")
        with library_write(db):
            result = stack_split_member_out(db, stack_id, image_key.strip())
        return jsonify(result), 200
    except StackMutationError as e:
        if e.status_code == 404:
            return error_not_found("stack")
        if e.status_code >= 500:
            return error_server_error(str(e))
        return error_bad_request(str(e))
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/<int:target_stack_id>/merge", methods=["POST"])
@with_db
@spec.validate(
    json=StackMergeRequest,
    resp=Response(
        HTTP_200=StackMergeResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-stacks'],
)
def post_stack_merge(db, target_stack_id: int):
    """Merge *source_stack_id* into *target_stack_id* (all members moved, source row deleted)."""
    try:
        if target_stack_id < 1:
            return error_not_found("stack")
        body = request.get_json(silent=True)
        if not body or not isinstance(body, dict):
            return error_bad_request("JSON body required")
        raw_source = body.get("source_stack_id")
        if raw_source is None:
            return error_bad_request("source_stack_id required")
        try:
            source_stack_id = int(raw_source)
        except (TypeError, ValueError):
            return error_bad_request("source_stack_id must be an integer")
        with library_write(db):
            result = stack_merge_into(db, target_stack_id, source_stack_id)
        return jsonify(result), 200
    except StackMutationError as e:
        if e.status_code == 404:
            return error_not_found("stack")
        if e.status_code >= 500:
            return error_server_error(str(e))
        return error_bad_request(str(e))
    except Exception as e:
        return error_server_error(str(e))


@stacks_bp.route("/<int:stack_id>/representative", methods=["POST"])
@with_db
@spec.validate(
    json=StackRepresentativeRequest,
    resp=Response(
        HTTP_200=StackRepresentativeResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-stacks'],
)
def post_stack_representative(db, stack_id: int):
    """Change which catalog key is the stack representative (must be a current member)."""
    try:
        if stack_id < 1:
            return error_not_found("stack")
        body = request.get_json(silent=True)
        if not body or not isinstance(body, dict):
            return error_bad_request("JSON body required")
        image_key = body.get("image_key")
        if not image_key or not isinstance(image_key, str):
            return error_bad_request("image_key required")
        with library_write(db):
            result = stack_set_representative(db, stack_id, image_key.strip())
        return jsonify(result), 200
    except StackMutationError as e:
        if e.status_code == 404:
            return error_not_found("stack")
        if e.status_code >= 500:
            return error_server_error(str(e))
        return error_bad_request(str(e))
    except Exception as e:
        return error_server_error(str(e))


__all__ = ("stacks_bp",)

File: apps/visualizer/backend/src/api/images/stacks.ts
/**
 * Burst stack members and mutations.
 *
 * Reuses catalog row shaping so stack members match the same image in the grid.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { ERROR_IMAGE_NOT_FOUND } from '../../constants/errors.js';
import { queryCatalogImagesByKeys } from '../../db/library/catalog-query.js';
import {
  listPendingStackSuggestions,
  rejectCatalogSimilarityPair,
  stackAcceptSuggestionPair,
} from '../../db/library/stack-suggestions.js';
import {
  listStackMemberKeys,
  stackExists,
  stackMergeInto,
  StackMutationError,
  stackSetRepresentative,
  stackSplitMemberOut,
} from '../../db/library/stacks.js';
import { libraryDb, type LibraryEnv } from '../library-db.js';
import { libraryWrite } from '../../db/library/write.js';
import { HttpError } from '../../utils/responses.js';
import { createOpenApiApp } from '../openapi.js';
import { jsonBody, withValidationError } from '../route-helpers.js';
import { ErrorBody } from '../schemas/errors.js';
import {
  StackMembersResponse,
  StackMergeRequest,
  StackMergeResponse,
  StackRepresentativeRequest,
  StackRepresentativeResponse,
  StackSplitMemberRequest,
  StackSplitMemberResponse,
  StackSuggestionAcceptResponse,
  StackSuggestionPairRequest,
  StackSuggestionRejectResponse,
  StackSuggestionsResponse,
} from '../schemas/stacks.js';
import { catalogThumbnailUrl, rowsToCatalogApiImages } from './row-shaping.js';

export const stacksRoutes = createOpenApiApp<LibraryEnv>();

stacksRoutes.use('/images/stacks/*', libraryDb({ writeForMethods: ['POST'] }));

/**
 * Stack pagination: limit capped at 100, default 20 — not the shared catalog helper
 * (500 / 50). Keep separate so a tidy-up does not merge the two.
 */
function clampStackPagination(
  limitRaw: string | undefined,
  offsetRaw: string | undefined,
): { limit: number; offset: number } {
  // Unparseable query values fall back to the default.
  const toInt = (raw: string | undefined, fallback: number): number =>
    raw !== undefined && /^\s*[+-]?\d+\s*$/.test(raw) ? Number.parseInt(raw.trim(), 10) : fallback;
  return {
    limit: Math.max(1, Math.min(toInt(limitRaw, 20), 100)),
    offset: Math.max(0, toInt(offsetRaw, 0)),
  };
}

/** A non-empty string; `""` is rejected even though it is schema-valid. */
function nonEmpty(value: string): string | null {
  return value.length > 0 ? value : null;
}

const stackIdParams = z.object({ stack_id: z.string() });

/**
 * Parse a `stack_id` path segment.
 *
 * A non-numeric id throws the legacy catch-all 400; an integer below 1 returns null
 * for the caller's 404. OpenAPI requires `{stack_id}` as a string, not an int converter.
 */
function parseStackId(raw: string): number | null {
  if (!/^\d+$/.test(raw)) {
    throw new HttpError(400, "invalid image_type; expected one of ('catalog', 'instagram')");
  }
  const n = Number.parseInt(raw, 10);
  return n >= 1 ? n : null;
}

// --- suggestions ------------------------------------------------------------

const suggestionsRoute = createRoute({
  method: 'get',
  path: '/images/stacks/suggestions',
  tags: ['images-stacks'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackSuggestionsResponse) },
  }),
});

stacksRoutes.openapi(suggestionsRoute, (c) => {
  const db = c.get('libraryDb');
  const { limit, offset } = clampStackPagination(c.req.query('limit'), c.req.query('offset'));
  const { rows, total } = listPendingStackSuggestions(db, { limit, offset });

  const items = [];
  for (const row of rows) {
    const keys = [String(row.seed_key), String(row.candidate_key)];
    const images = rowsToCatalogApiImages(queryCatalogImagesByKeys(db, keys));
    const byKey = new Map(images.map((img) => [String(img.key), img]));
    const imageA = byKey.get(String(row.seed_key));
    const imageB = byKey.get(String(row.candidate_key));
    // Either end being collapsed out of the primary grid drops the pair: a
    // suggestion you cannot see both halves of is not reviewable.
    if (!imageA || !imageB) continue;

    imageA.thumbnail_url = catalogThumbnailUrl(String(imageA.key));
    imageB.thumbnail_url = catalogThumbnailUrl(String(imageB.key));
    items.push({
      group_id: Math.trunc(row.group_id),
      image_a: imageA,
      image_b: imageB,
      similarity: Number(row.similarity ?? 0),
      why_matched: String(row.why_matched ?? ''),
      time_gap_seconds:
        row.time_gap_seconds === null || row.time_gap_seconds === undefined
          ? null
          : Math.trunc(row.time_gap_seconds),
    });
  }
  return c.json({ items, total }, 200);
});

const acceptRoute = createRoute({
  method: 'post',
  path: '/images/stacks/suggestions/accept',
  tags: ['images-stacks'],
  request: { body: { content: jsonBody(StackSuggestionPairRequest) } },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackSuggestionAcceptResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(acceptRoute, (c) => {
  const db = c.get('libraryDb');
  const { image_key_a: rawA, image_key_b: rawB } = c.req.valid('json');
  const keyA = nonEmpty(rawA);
  const keyB = nonEmpty(rawB);
  if (!keyA || !keyB) return c.json({ error: 'image_key_a and image_key_b required' }, 400);

  try {
    const result = libraryWrite(db, () => stackAcceptSuggestionPair(db, keyA.trim(), keyB.trim()));
    // The merge branch also returns `merged_stack_id`, which this response model
    // forbids — narrow rather than spread.
    return c.json({ stack: result.stack }, 200);
  } catch (e) {
    if (e instanceof StackMutationError) {
      // This route reports a missing *image*, unlike the stack-scoped routes.
      if (e.statusCode === 404) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);
      if (e.statusCode < 500) return c.json({ error: e.message }, 400);
    }
    throw e;
  }
});

const rejectRoute = createRoute({
  method: 'post',
  path: '/images/stacks/suggestions/reject',
  tags: ['images-stacks'],
  request: { body: { content: jsonBody(StackSuggestionPairRequest) } },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackSuggestionRejectResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(rejectRoute, (c) => {
  const db = c.get('libraryDb');
  const { image_key_a: rawA, image_key_b: rawB } = c.req.valid('json');
  const keyA = nonEmpty(rawA);
  const keyB = nonEmpty(rawB);
  if (!keyA || !keyB) return c.json({ error: 'image_key_a and image_key_b required' }, 400);

  const a = keyA.trim();
  const b = keyB.trim();
  if (a === b) return c.json({ error: 'image_key_a and image_key_b must differ' }, 400);

  libraryWrite(db, () => rejectCatalogSimilarityPair(db, a, b));
  return c.json({ image_key_a: a, image_key_b: b, rejected: true }, 200);
});

// --- members ----------------------------------------------------------------

const membersRoute = createRoute({
  method: 'get',
  path: '/images/stacks/{stack_id}/members',
  tags: ['images-stacks'],
  request: { params: stackIdParams },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackMembersResponse) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(membersRoute, (c) => {
  const db = c.get('libraryDb');
  const stackId = parseStackId(c.req.param('stack_id'));
  if (stackId === null) return c.json({ error: 'stack not found' }, 404);
  if (!stackExists(db, stackId)) return c.json({ error: 'stack not found' }, 404);

  const keys = listStackMemberKeys(db, stackId);
  if (keys.length === 0) return c.json({ items: [] }, 200);

  // `primaryGridOnly: false` is the point of this query — the reason to open a stack
  // is to see the members the grid collapses away.
  const items = rowsToCatalogApiImages(
    queryCatalogImagesByKeys(db, keys, { scorePerspective: null, primaryGridOnly: false }),
  );
  for (const it of items) it.thumbnail_url = catalogThumbnailUrl(String(it.key));
  return c.json({ items }, 200);
});

// --- mutations --------------------------------------------------------------

/**
 * Stack mutation error mapping is repeated inline: a helper returning `Response`
 * erases the typed-response information `@hono/zod-openapi` checks against.
 */

const splitMemberRoute = createRoute({
  method: 'post',
  path: '/images/stacks/{stack_id}/split-member',
  tags: ['images-stacks'],
  request: { params: stackIdParams, body: { content: jsonBody(StackSplitMemberRequest) } },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackSplitMemberResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(splitMemberRoute, (c) => {
  const db = c.get('libraryDb');
  const stackId = parseStackId(c.req.param('stack_id'));
  if (stackId === null) return c.json({ error: 'stack not found' }, 404);

  const imageKey = nonEmpty(c.req.valid('json').image_key);
  if (!imageKey) return c.json({ error: 'image_key required' }, 400);

  try {
    return c.json(libraryWrite(db, () => stackSplitMemberOut(db, stackId, imageKey.trim())), 200);
  } catch (e) {
    if (e instanceof StackMutationError) {
      if (e.statusCode === 404) return c.json({ error: 'stack not found' }, 404);
      if (e.statusCode < 500) return c.json({ error: e.message }, 400);
    }
    throw e;
  }
});

const mergeRoute = createRoute({
  method: 'post',
  path: '/images/stacks/{target_stack_id}/merge',
  tags: ['images-stacks'],
  request: {
    params: z.object({ target_stack_id: z.string() }),
    body: { content: jsonBody(StackMergeRequest) },
  },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackMergeResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(mergeRoute, (c) => {
  const db = c.get('libraryDb');
  const targetStackId = parseStackId(c.req.param('target_stack_id'));
  if (targetStackId === null) return c.json({ error: 'stack not found' }, 404);

  const sourceStackId = c.req.valid('json').source_stack_id;
  try {
    return c.json(libraryWrite(db, () => stackMergeInto(db, targetStackId, sourceStackId)), 200);
  } catch (e) {
    if (e instanceof StackMutationError) {
      if (e.statusCode === 404) return c.json({ error: 'stack not found' }, 404);
      if (e.statusCode < 500) return c.json({ error: e.message }, 400);
    }
    throw e;
  }
});

const representativeRoute = createRoute({
  method: 'post',
  path: '/images/stacks/{stack_id}/representative',
  tags: ['images-stacks'],
  request: { params: stackIdParams, body: { content: jsonBody(StackRepresentativeRequest) } },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(StackRepresentativeResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

stacksRoutes.openapi(representativeRoute, (c) => {
  const db = c.get('libraryDb');
  const stackId = parseStackId(c.req.param('stack_id'));
  if (stackId === null) return c.json({ error: 'stack not found' }, 404);

  const imageKey = nonEmpty(c.req.valid('json').image_key);
  if (!imageKey) return c.json({ error: 'image_key required' }, 400);

  try {
    return c.json(libraryWrite(db, () => stackSetRepresentative(db, stackId, imageKey.trim())), 200);
  } catch (e) {
    if (e instanceof StackMutationError) {
      if (e.statusCode === 404) return c.json({ error: 'stack not found' }, 404);
      if (e.statusCode < 500) return c.json({ error: e.message }, 400);
    }
    throw e;
  }
});
